sched/task/exit.c

```diff
@@ -70,7 +68,7 @@ void _exit(int status)
  *
  ****************************************************************************/
 
-void exit(int status)
+void nx_exit(int status)
```

Comment:
Do we really need nx_exit? only _exit is really needed, I think.

1. exit in libc call register exit function and flush/close file stream
2. Continue the termination process by call _exit
3. _exit in kernel release the rest resource

So, we just need change exit at line 73 to _exit and remove line 101 finally.

Reply:
Calling _exit() directly causes a full system crash. The problem with _exit() is that it removes the current task from the ready-to-run list, and any call to this_task() after that crashes as it does not point to the task that is in execution. File cleanup uses a semaphore for instance and there we have a test for this_task() != idle task, which fails because the running task is not what this_task() returns.

If you look at the the old implementation of exit() in kernel side does file stream clean up FIRST, and then calls _exit(), this works because the rtr list head still points to the correct task.

Reply:
If so, we can reorder the code in _exit to correct the sequence. It can't resolve by introducing nx_exit. So, I would suggest that:

1. Remove nx_exit
2. Make exit as a normal libc function
3. Make _exit as the kernel function

atexit related function look in a good shape, the next step is simple now:

1. Create a new patch to remove the kernel side atexit and enable libc's version
2. Remove the exit callback from _exit

The last step is moving file stream flush/close to the userspace. The last work is my TODO list, actually I have a draft patch in my local branch for it.

Reply:
If so, we can reorder the code in _exit to correct the sequence. It can't resolve by introducing nx_exit. So, I would suggest that:

Remove nx_exit
Make exit as a normal libc function
Make _exit as the kernel function
atexit related function look in a good shape, the next step is simple now:

Create a new patch to remove the kernel side atexit and enable libc's version
Remove the exit callback from _exit
The last step is moving file stream flush/close to the userspace. The last work is my TODO list, actually I have a draft patch in my local branch for it.

Reply:
Ok, I'll add the patches that remove the kernel side callbacks to this PR. I agree the file stream handling should be in userland too, but that can be a separate PR. If you have a draft maybe you can then merge that on top of this?

Reply:
Sure, I will prepare that patch once your patch get merged.